File: S1/Reseaux/TPs/TP3/src/src/EchoClient.java
/*  echo / client simple
    Master Informatique 2012 -- Université Aix-Marseille
    Emmanuel Godard - Bilel Derbel
*/

import java.io.*;
import java.net.*;

public class EchoClient {

  public static void main(String[] args) throws IOException {
    Socket echoSocket; // la socket client
    String ip; // adresse IPv4 du serveur en notation pointée
    int port; // port TCP serveur
    BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
    PrintWriter out;
    BufferedReader in;
    boolean fini = false;

    /* Traitement des arguments */
    if (args.length != 2) {
        /* erreur de syntaxe */
      System.out.println("Usage: java EchoClient @server @port");
      System.exit(1);
    }
    ip = args[0];
    port = Integer.parseInt(args[1]);

    if (port > 65535) {
      System.err.println("Port hors limite");
      System.exit(3);
    }

    /* Connexion */
    System.out.println("Essai de connexion à  " + ip + " sur le port " + port + "\n");
    try {
      echoSocket = new Socket(ip, port);
      System.err.println("le n° de la socket est : " + echoSocket);
      /* Initialisation d'agréables flux d'entrée/sortie */
      out = new PrintWriter(echoSocket.getOutputStream(), true);
      in = new BufferedReader(new InputStreamReader(echoSocket.getInputStream()));
    } catch (UnknownHostException e) {
      System.err.println("Connexion: hôte inconnu : " + ip);
      e.printStackTrace();
      return;
    }

    /* Session */
    try {
      while (true) {
        /* Jusqu'à fermeture de la socket (ou de stdin)     */
        /* recopier à l'écran ce qui est lu dans la socket  */
        /* recopier dans la socket ce qui est lu dans stdin */

        String lu;
        String tampon;

        /* réception des données */
        lu = in.readLine();
        if (lu == null) {
          System.err.println("Connexion terminée par l'hôte distant");
          break; /* on sort de la boucle infinie */
        }
        System.out.println("reçu: " + lu);

        if (fini == true) break; /* on sort de la boucle infinie */

        /* recopier dans la socket ce qui est entré au clavier */
        tampon = stdin.readLine();
        if (tampon == null) {
          fini = true;
          System.err.println("Connexion terminée !!");
          System.err.println("Hôte distant informé...");
          echoSocket.shutdownOutput(); /* terminaison explicite de la socket
                                          dans le sens client -> serveur */
          /* On ne sort pas de la boucle tout de suite ... */
        } else {
            /* envoi des données */
          out.println(tampon);
        }
      }

      /* On ferme tout */
      in.close();
      out.close();
      stdin.close();
      echoSocket.close();

      System.err.println("Fin de la session.");
    } catch (IOException e) {
      System.err.println("Erreur E/S socket");
      e.printStackTrace();
      System.exit(8);
    }

    return;
  }
}

File: S1/Reseaux/TPs/TP3/src/src/echoclient.c
/*  echo / client simple
    Master Informatique 2012 -- Université Aix-Marseille  
    Emmanuel Godard
*/

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#define MAXLIGNE 64

int main(int argc, char *argv[])
{
  char * hote; /* nom d'hôte du  serveur */   
  char * port; /* port TCP du serveur */   
  char ip[NI_MAXHOST]; /* adresse IPv4 en notation pointée */
  struct addrinfo *resol; /* struct pour la résolution de nom */
  int s; /* descripteur de socket */

  /* Traitement des arguments */
  if (argc!=3) {/* erreur de syntaxe */
    printf("Usage: %s hote port\n",argv[0]);
    exit(1);
  }
  hote=argv[1]; /* nom d'hôte du  serveur */   
  port=argv[2]; /* port TCP du serveur */   

  /* Résolution de l'hôte */
  if ( getaddrinfo(hote,port,NULL, &resol) < 0 ){
    perror("résolution adresse");
    exit(2);
  }

  /* On extrait l'addresse IP */
  sprintf(ip,"%s",inet_ntoa(((struct sockaddr_in*)resol->ai_addr)->sin_addr));

  /* Création de la socket, de type TCP / IP */
  /* On ne considère que la première adresse renvoyée par getaddrinfo */
  if ((s=socket(resol->ai_family,resol->ai_socktype, resol->ai_protocol))<0) {
    perror("allocation de socket");
    exit(3);
  }
  fprintf(stderr,"le n° de la socket est : %i\n",s);

  /* Connexion */
  fprintf(stderr,"Essai de connexion à %s (%s) sur le port %s\n\n",
	  hote,ip,port);
  if (connect(s,resol->ai_addr,sizeof(struct sockaddr_in))<0) {
    perror("connexion");
    exit(4);
  }
  freeaddrinfo(resol); /* /!\ Libération mémoire */

  /* Session */
  char tampon[MAXLIGNE + 3]; /* tampons pour les communications */
  ssize_t lu;
  int fini=0;
  while( 1 ) { 
    /* Jusqu'à fermeture de la socket (ou de stdin)     */
    /* recopier à l'écran ce qui est lu dans la socket  */
    /* recopier dans la socket ce qui est lu dans stdin */

    /* réception des données */
    lu = recv(s,tampon,MAXLIGNE,0); /* bloquant */
    if (lu == 0 ) {
      fprintf(stderr,"Connexion terminée par l'hôte distant\n");
      break; /* On sort de la boucle infinie */
    }
    tampon[lu] = '\0';
    printf("reçu: %s",tampon);
    if ( fini == 1 )
      break;  /* on sort de la boucle infinie*/
    
    /* recopier dans la socket ce qui est entré au clavier */    
    if ( fgets(tampon,MAXLIGNE - 2,stdin) == NULL ){/* entrée standard fermée */
      fini=1;
      fprintf(stderr,"Connexion terminée !!\n");
      fprintf(stderr,"Hôte distant informé...\n");
      shutdown(s, SHUT_WR); /* terminaison explicite de la socket 
			     dans le sens client -> serveur */
      /* On ne sort pas de la boucle tout de suite ... */
    }else{   /* envoi des données */
      send(s,tampon,strlen(tampon),0);
    }
  } 
  /* Destruction de la socket */
  close(s);

  fprintf(stderr,"Fin de la session.\n");
  return EXIT_SUCCESS;
}
